File: src/worktrees.ts
/**
 * Helpers for `git worktree list --porcelain` parsing, under-cwd path mapping,
 * and merge classification. Path mapping may touch the filesystem (realpath/stat).
 */

import * as fs from 'fs';
import * as path from 'path';
import { DETACHED_HEAD_BRANCH } from './helpers.js';

/**
 * One worktree from `git worktree list --porcelain`.
 */
export interface GitWorktreeListEntry {
  /** Absolute worktree path. */
  path: string;
  /** HEAD commit hash when present. */
  head?: string;
  /** Local branch name (no `refs/heads/` prefix) when checked out on a branch. */
  branch?: string;
  /** True when the worktree is bare. */
  bare: boolean;
  /** True when HEAD is detached. */
  detached: boolean;
}

/**
 * Device + inode identity for bind-mount / hard-link equivalence.
 */
export interface PathIdentity {
  dev: number;
  ino: number;
}

/**
 * Filesystem ops used by `linkedWorktreesUnderCwd` (injectable for tests).
 */
export interface LinkedWorktreePathIo {
  /** Resolve symlinks; return input path when the target is missing. */
  realpath: (absPath: string) => string;
  /** Stat identity, or null when the path does not exist. */
  identity: (absPath: string) => PathIdentity | null;
}

const defaultLinkedWorktreePathIo: LinkedWorktreePathIo = {
  realpath: (absPath) => {
    try {
      return fs.realpathSync(absPath);
    } catch {
      return absPath;
    }
  },
  identity: (absPath) => {
    try {
      const st = fs.statSync(absPath);
      return { dev: st.dev, ino: st.ino };
    } catch {
      return null;
    }
  },
};

function sameIdentity(a: PathIdentity | null, b: PathIdentity | null): boolean {
  return a !== null && b !== null && a.dev === b.dev && a.ino === b.ino;
}

function underDir(abs: string, dir: string): boolean {
  return abs === dir || abs.startsWith(dir + path.sep);
}

function posixRel(from: string, to: string): string {
  return path.relative(from, to).split(path.sep).join('/');
}

/** Join workspace-relative segments, skipping empties (primary may equal cwd). */
function joinWorkspaceRel(...parts: string[]): string {
  return parts.filter((p) => p.length > 0).join('/');
}

/**
 * Remap via primary inode when the registered path is a bind-mount alias.
 * Returns null when the entry is not under that primary identity.
 */
function remapViaPrimaryIdentity(
  abs: string,
  cwd: string,
  primary: string,
  primaryId: PathIdentity,
  io: LinkedWorktreePathIo,
): { absPath: string; relPath: string } | null {
  const primaryRel = posixRel(cwd, primary);
  const suffix: string[] = [];
  let cur = abs;
  for (;;) {
    if (sameIdentity(io.identity(cur), primaryId) || cur === primary) {
      const relPath = joinWorkspaceRel(primaryRel, [...suffix].reverse().join('/'));
      return {
        absPath: relPath ? path.resolve(cwd, ...relPath.split('/')) : path.resolve(cwd),
        relPath,
      };
    }
    const parent = path.dirname(cur);
    if (parent === cur) break;
    suffix.push(path.basename(cur));
    cur = parent;
  }
  return null;
}

/**
 * Map a linked worktree absolute path to a workspace-relative path under `cwd`,
 * or null when it is outside the workspace.
 *
 * Prefer the cwd-visible primary prefix: paths already under that primary use
 * string relative after realpath; bind-mount aliases (same inode, different
 * absolute prefix) remapped via inode walk even when the alias also sits under
 * cwd. Other under-cwd paths keep a plain relative mapping.
 */
export function mapLinkedWorktreeRelPath(
  entryAbs: string,
  cwdAbs: string,
  primaryAbs: string,
  io: LinkedWorktreePathIo = defaultLinkedWorktreePathIo,
): { absPath: string; relPath: string } | null {
  const cwd = io.realpath(path.resolve(cwdAbs));
  const primary = io.realpath(path.resolve(primaryAbs));
  const abs = io.realpath(path.resolve(entryAbs));
  const primaryId = io.identity(primary);

  if (sameIdentity(io.identity(abs), primaryId) || abs === primary) {
    return null;
  }

  if (!underDir(primary, cwd)) {
    return null;
  }

  // Already under the cwd-visible primary — no remapping needed.
  if (underDir(abs, primary)) {
    return {
      absPath: abs,
      relPath: posixRel(cwd, abs),
    };
  }

  // Bind-mount alias of a path under this primary (may still be under cwd).
  if (primaryId) {
    const remapped = remapViaPrimaryIdentity(abs, cwd, primary, primaryId, io);
    if (remapped) return remapped;
  }

  if (underDir(abs, cwd)) {
    return {
      absPath: abs,
      relPath: posixRel(cwd, abs),
    };
  }

  return null;
}

/**
 * Listed worktree path git will accept: exact string, else same inode.
 */
function listedWorktreePath(
  entries: GitWorktreeListEntry[],
  worktreeAbs: string,
  io: LinkedWorktreePathIo,
): string {
  const wt = path.resolve(worktreeAbs);
  for (const entry of entries) {
    if (entry.bare) continue;
    if (path.resolve(entry.path) === wt) return wt;
  }
  const wtId = io.identity(wt);
  if (!wtId) return wt;
  for (const entry of entries) {
    if (entry.bare) continue;
    if (sameIdentity(io.identity(path.resolve(entry.path)), wtId)) {
      return path.resolve(entry.path);
    }
  }
  return wt;
}

/**
 * Ancestor of `abs` with the same inode as `primaryAbs` (bind-mount alias
 * of the primary). Null when no ancestor matches.
 */
function registeredPrimaryAbs(
  abs: string,
  primaryAbs: string,
  io: LinkedWorktreePathIo,
): string | null {
  const primaryId = io.identity(path.resolve(primaryAbs));
  if (!primaryId) return null;
  let cur = path.resolve(abs);
  for (;;) {
    if (sameIdentity(io.identity(cur), primaryId)) return cur;
    const parent = path.dirname(cur);
    if (parent === cur) return null;
    cur = parent;
  }
}

/**
 * Paths git will accept for `worktree remove` when the TUI path is a
 * bind-mount alias of the registered worktree.
 *
 * `gitPath` is the porcelain `worktree` line (inode match when prefixes
 * differ). `gitCwd` is the registered primary prefix — the ancestor of
 * `gitPath` with the same inode as `primaryAbs` — so gitdir back-pointers
 * compare equal. Callers fall back to the TUI paths when nothing matches.
 */
export function resolveWorktreeRemoveTarget(
  entries: GitWorktreeListEntry[],
  primaryAbs: string,
  worktreeAbs: string,
  io: LinkedWorktreePathIo = defaultLinkedWorktreePathIo,
): { gitCwd: string; gitPath: string } {
  const gitPath = listedWorktreePath(entries, worktreeAbs, io);
  const gitCwd = registeredPrimaryAbs(gitPath, primaryAbs, io) ?? path.resolve(primaryAbs);
  return { gitCwd, gitPath };
}

/**
 * Parse `git worktree list --porcelain` stdout into entries.
 *
 * Strips `refs/heads/` from branch lines. Paths are normalized with `path.resolve`.
 */
export function parseWorktreeListPorcelain(text: string): GitWorktreeListEntry[] {
  const entries: GitWorktreeListEntry[] = [];
  let current: GitWorktreeListEntry | null = null;

  const flush = () => {
    if (current) {
      entries.push(current);
      current = null;
    }
  };

  for (const rawLine of text.split('\n')) {
    const line = rawLine.trimEnd();
    if (!line) {
      flush();
      continue;
    }
    if (line.startsWith('worktree ')) {
      flush();
      current = {
        path: path.resolve(line.slice('worktree '.length)),
        bare: false,
        detached: false,
      };
      continue;
    }
    if (!current) continue;
    if (line.startsWith('HEAD ')) {
      current.head = line.slice('HEAD '.length);
      continue;
    }
    if (line.startsWith('branch ')) {
      const ref = line.slice('branch '.length);
      current.branch = ref.startsWith('refs/heads/') ? ref.slice('refs/heads/'.length) : ref;
      continue;
    }
    if (line === 'bare') {
      current.bare = true;
      continue;
    }
    if (line === 'detached') {
      current.detached = true;
    }
  }
  flush();
  return entries;
}

/**
 * Linked (non-primary, non-bare) worktrees whose paths fall under `cwdAbs`.
 *
 * Uses realpath for symlink layouts. Bind-mount aliases of the primary (same
 * inode, different absolute prefix) are remapped to the cwd-visible relative path.
 *
 * `relPath` is relative to `cwdAbs` with `/` separators.
 */
export function linkedWorktreesUnderCwd(
  entries: GitWorktreeListEntry[],
  cwdAbs: string,
  primaryAbs: string,
  io: LinkedWorktreePathIo = defaultLinkedWorktreePathIo,
): { absPath: string; relPath: string }[] {
  const out: { absPath: string; relPath: string }[] = [];
  for (const entry of entries) {
    if (entry.bare) continue;
    const mapped = mapLinkedWorktreeRelPath(entry.path, cwdAbs, primaryAbs, io);
    if (!mapped) continue;
    out.push(mapped);
  }
  return out;
}

/**
 * Whether a branch tip is merged into the default branch, or unknown.
 *
 * - `null` when on the default branch, or when ancestry is unknown
 * - otherwise the `isAncestorOfDefault` boolean
 */
export type MergedIntoDefault = boolean | null;

/**
 * Classify merge-into-default from branch name + ancestry probe result.
 */
export function classifyMergedIntoDefault(args: {
  branch: string;
  defaultBranch: string;
  isAncestorOfDefault: boolean | null;
}): MergedIntoDefault {
  if (args.branch === args.defaultBranch) return null;
  // Detached HEAD is not a branch tip relative to default — omit merge marks.
  if (args.branch === DETACHED_HEAD_BRANCH) return null;
  if (args.isAncestorOfDefault === null) return null;
  return args.isAncestorOfDefault;
}

File: crates/workspace-status/src/worktrees.rs
//! Worktree porcelain parse, under-cwd mapping, merge classification.

use std::fs;
use std::path::{Path, PathBuf};

use crate::helpers::DETACHED_HEAD_BRANCH;

#[derive(Debug, Clone)]
pub struct GitWorktreeListEntry {
    pub path: PathBuf,
    pub head: Option<String>,
    pub branch: Option<String>,
    pub bare: bool,
    pub detached: bool,
}

/// Filesystem identity used to match a TUI path to a git-registered worktree.
///
/// Unix uses `(dev, ino)` so bind-mount aliases match. Other platforms have no
/// inodes; identity is canonical path plus size and mtime.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PathIdentity {
    #[cfg(unix)]
    pub dev: u64,
    #[cfg(unix)]
    pub ino: u64,
    #[cfg(not(unix))]
    path: PathBuf,
    #[cfg(not(unix))]
    len: u64,
    #[cfg(not(unix))]
    mtime: Option<std::time::SystemTime>,
}

fn realpath(abs: &Path) -> PathBuf {
    fs::canonicalize(abs).unwrap_or_else(|_| abs.to_path_buf())
}

fn identity(abs: &Path) -> Option<PathIdentity> {
    let st = fs::metadata(abs).ok()?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::MetadataExt;
        Some(PathIdentity {
            dev: st.dev(),
            ino: st.ino(),
        })
    }
    #[cfg(not(unix))]
    {
        Some(PathIdentity {
            path: abs.to_path_buf(),
            len: st.len(),
            mtime: st.modified().ok(),
        })
    }
}

fn same_identity(a: Option<&PathIdentity>, b: Option<&PathIdentity>) -> bool {
    matches!((a, b), (Some(x), Some(y)) if x == y)
}

fn under_dir(abs: &Path, dir: &Path) -> bool {
    abs == dir || abs.starts_with(dir)
}

fn posix_rel(from: &Path, to: &Path) -> String {
    pathdiff_rel(from, to)
}

fn pathdiff_rel(from: &Path, to: &Path) -> String {
    match to.strip_prefix(from) {
        Ok(rel) => rel.to_string_lossy().replace('\\', "/"),
        Err(_) => to.to_string_lossy().replace('\\', "/"),
    }
}

fn join_workspace_rel(parts: &[&str]) -> String {
    parts
        .iter()
        .filter(|p| !p.is_empty())
        .copied()
        .collect::<Vec<_>>()
        .join("/")
}

fn remap_via_primary_identity(
    abs: &Path,
    cwd: &Path,
    primary: &Path,
    primary_id: PathIdentity,
) -> Option<(PathBuf, String)> {
    let primary_rel = posix_rel(cwd, primary);
    let mut suffix: Vec<String> = Vec::new();
    let mut cur = abs.to_path_buf();
    loop {
        if same_identity(identity(&cur).as_ref(), Some(&primary_id)) || cur == primary {
            suffix.reverse();
            let extra = suffix.join("/");
            let rel_path = join_workspace_rel(&[primary_rel.as_str(), extra.as_str()]);
            let abs_path = if rel_path.is_empty() {
                cwd.to_path_buf()
            } else {
                cwd.join(PathBuf::from(
                    rel_path.replace('/', std::path::MAIN_SEPARATOR_STR),
                ))
            };
            return Some((abs_path, rel_path));
        }
        match cur.parent() {
            Some(parent) if parent != cur => {
                if let Some(name) = cur.file_name() {
                    suffix.push(name.to_string_lossy().into_owned());
                }
                cur = parent.to_path_buf();
            }
            _ => break,
        }
    }
    None
}

/// Main checkout: `.git` is a directory. Linked extras use a gitfile.
pub fn is_main_worktree_checkout(dir: &Path) -> bool {
    fs::metadata(dir.join(".git"))
        .map(|st| st.is_dir())
        .unwrap_or(false)
}

/// Linked extra: `.git` is a file (`gitdir: …`), not the main checkout.
pub fn is_linked_worktree_checkout(dir: &Path) -> bool {
    fs::metadata(dir.join(".git"))
        .map(|st| st.is_file())
        .unwrap_or(false)
}

/// Map a linked worktree absolute path to a workspace-relative path under cwd.
pub fn map_linked_worktree_rel_path(
    entry_abs: &Path,
    cwd_abs: &Path,
    primary_abs: &Path,
) -> Option<(PathBuf, String)> {
    let cwd = realpath(cwd_abs);
    let primary = realpath(primary_abs);
    let abs = realpath(entry_abs);
    let primary_id = identity(&primary);

    if same_identity(identity(&abs).as_ref(), primary_id.as_ref()) || abs == primary {
        return None;
    }
    if !under_dir(&primary, &cwd) {
        return None;
    }
    if under_dir(&abs, &primary) {
        return Some((abs.clone(), posix_rel(&cwd, &abs)));
    }
    if let Some(id) = primary_id {
        if let Some(remapped) = remap_via_primary_identity(&abs, &cwd, &primary, id) {
            return Some(remapped);
        }
    }
    if under_dir(&abs, &cwd) {
        return Some((abs.clone(), posix_rel(&cwd, &abs)));
    }
    None
}

pub fn parse_worktree_list_porcelain(text: &str) -> Vec<GitWorktreeListEntry> {
    let mut entries = Vec::new();
    let mut current: Option<GitWorktreeListEntry> = None;

    let flush = |current: &mut Option<GitWorktreeListEntry>,
                 entries: &mut Vec<GitWorktreeListEntry>| {
        if let Some(entry) = current.take() {
            entries.push(entry);
        }
    };

    for raw in text.split('\n') {
        let line = raw.trim_end();
        if line.is_empty() {
            flush(&mut current, &mut entries);
            continue;
        }
        if let Some(rest) = line.strip_prefix("worktree ") {
            flush(&mut current, &mut entries);
            current = Some(GitWorktreeListEntry {
                path: PathBuf::from(rest),
                head: None,
                branch: None,
                bare: false,
                detached: false,
            });
            continue;
        }
        let Some(cur) = current.as_mut() else {
            continue;
        };
        if let Some(rest) = line.strip_prefix("HEAD ") {
            cur.head = Some(rest.to_string());
            continue;
        }
        if let Some(rest) = line.strip_prefix("branch ") {
            cur.branch = Some(rest.strip_prefix("refs/heads/").unwrap_or(rest).to_string());
            continue;
        }
        if line == "bare" {
            cur.bare = true;
            continue;
        }
        if line == "detached" {
            cur.detached = true;
        }
    }
    flush(&mut current, &mut entries);
    entries
}

pub fn linked_worktrees_under_cwd(
    entries: &[GitWorktreeListEntry],
    cwd_abs: &Path,
    primary_abs: &Path,
) -> Vec<(PathBuf, String)> {
    let mut out = Vec::new();
    for entry in entries {
        if entry.bare {
            continue;
        }
        if let Some(mapped) = map_linked_worktree_rel_path(&entry.path, cwd_abs, primary_abs) {
            out.push(mapped);
        }
    }
    out
}

/// Paths git accepts for `worktree remove` when the TUI path is a bind-mount alias.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WorktreeRemoveTarget {
    pub git_cwd: PathBuf,
    pub git_path: PathBuf,
}

fn resolve_abs(path: &Path) -> PathBuf {
    if path.is_absolute() {
        path.to_path_buf()
    } else {
        std::env::current_dir()
            .unwrap_or_else(|_| PathBuf::from("."))
            .join(path)
    }
}

fn listed_worktree_path(entries: &[GitWorktreeListEntry], worktree_abs: &Path) -> PathBuf {
    let wt = resolve_abs(worktree_abs);
    for entry in entries {
        if entry.bare {
            continue;
        }
        if resolve_abs(&entry.path) == wt {
            return wt;
        }
    }
    let wt_id = identity(&wt);
    if wt_id.is_none() {
        return wt;
    }
    for entry in entries {
        if entry.bare {
            continue;
        }
        let listed = resolve_abs(&entry.path);
        if same_identity(identity(&listed).as_ref(), wt_id.as_ref()) {
            return listed;
        }
    }
    wt
}

fn registered_primary_abs(abs: &Path, primary_abs: &Path) -> Option<PathBuf> {
    let primary_id = identity(&resolve_abs(primary_abs))?;
    let mut cur = resolve_abs(abs);
    loop {
        if same_identity(identity(&cur).as_ref(), Some(&primary_id)) {
            return Some(cur);
        }
        match cur.parent() {
            Some(parent) if parent != cur => cur = parent.to_path_buf(),
            _ => return None,
        }
    }
}

/// `gitPath` is the porcelain worktree line (inode match on Unix when prefixes
/// differ; path + size + mtime elsewhere).
/// `gitCwd` is the registered primary prefix so gitdir back-pointers match.
pub fn resolve_worktree_remove_target(
    entries: &[GitWorktreeListEntry],
    primary_abs: &Path,
    worktree_abs: &Path,
) -> WorktreeRemoveTarget {
    let git_path = listed_worktree_path(entries, worktree_abs);
    let git_cwd =
        registered_primary_abs(&git_path, primary_abs).unwrap_or_else(|| resolve_abs(primary_abs));
    WorktreeRemoveTarget { git_cwd, git_path }
}

pub fn classify_merged_into_default(
    branch: &str,
    default_branch: &str,
    is_ancestor_of_default: Option<bool>,
) -> Option<bool> {
    if branch == default_branch {
        return None;
    }
    if branch == DETACHED_HEAD_BRANCH {
        return None;
    }
    is_ancestor_of_default
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::Path;

    #[test]
    fn parse_porcelain_branch_and_detached() {
        let text = "\
worktree /tmp/app
HEAD abc
branch refs/heads/main

worktree /tmp/app/.worktrees/feat
HEAD def
detached
";
        let entries = parse_worktree_list_porcelain(text);
        assert_eq!(entries.len(), 2);
        assert_eq!(entries[0].branch.as_deref(), Some("main"));
        assert!(entries[1].detached);
    }

    #[test]
    fn classify_merge_skips_default_and_detached() {
        assert_eq!(
            classify_merged_into_default("main", "main", Some(true)),
            None
        );
        assert_eq!(
            classify_merged_into_default(DETACHED_HEAD_BRANCH, "main", Some(true)),
            None
        );
        assert_eq!(
            classify_merged_into_default("feature/x", "main", Some(true)),
            Some(true)
        );
    }

    #[test]
    fn resolve_remove_target_uses_listed_path() {
        let entries = vec![
            GitWorktreeListEntry {
                path: PathBuf::from("/tmp/app"),
                head: None,
                branch: Some("main".into()),
                bare: false,
                detached: false,
            },
            GitWorktreeListEntry {
                path: PathBuf::from("/tmp/app/.worktrees/feat"),
                head: None,
                branch: Some("feature/x".into()),
                bare: false,
                detached: false,
            },
        ];
        let target = resolve_worktree_remove_target(
            &entries,
            Path::new("/tmp/app"),
            Path::new("/tmp/app/.worktrees/feat"),
        );
        assert_eq!(target.git_path, PathBuf::from("/tmp/app/.worktrees/feat"));
        assert_eq!(target.git_cwd, PathBuf::from("/tmp/app"));
    }

    #[test]
    fn resolve_remove_target_falls_back_when_unlisted() {
        let target = resolve_worktree_remove_target(
            &[],
            Path::new("/tmp/app"),
            Path::new("/tmp/app/.worktrees/feat"),
        );
        assert_eq!(target.git_path, PathBuf::from("/tmp/app/.worktrees/feat"));
        assert_eq!(target.git_cwd, PathBuf::from("/tmp/app"));
    }

    #[test]
    fn identity_same_path_matches_and_parent_differs() {
        let dir = std::env::temp_dir();
        let a = identity(&dir);
        let b = identity(&dir);
        assert!(a.is_some());
        assert!(same_identity(a.as_ref(), b.as_ref()));
        if let Some(parent) = dir.parent() {
            if parent != dir {
                let parent_id = identity(parent);
                if parent_id.is_some() {
                    assert!(!same_identity(a.as_ref(), parent_id.as_ref()));
                }
            }
        }
    }

    #[test]
    fn gitfile_vs_gitdir_classifies_linked_vs_main() {
        let root = std::env::temp_dir().join(format!(
            "ws-wt-kind-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));
        let main = root.join("main");
        let linked = root.join("linked");
        fs::create_dir_all(&main).unwrap();
        fs::create_dir_all(&linked).unwrap();
        fs::create_dir_all(main.join(".git")).unwrap();
        fs::write(linked.join(".git"), "gitdir: ../main/.git/worktrees/x\n").unwrap();
        assert!(is_main_worktree_checkout(&main));
        assert!(!is_linked_worktree_checkout(&main));
        assert!(is_linked_worktree_checkout(&linked));
        assert!(!is_main_worktree_checkout(&linked));
        let _ = fs::remove_dir_all(&root);
    }
}
